### ISSUE TITLE
whitelist for port in addition to port range for `NodePort` services

### ISSUE THREAD

**Issue by yquansah** [kubernetes/kubernetes#127934]
### What would you like to be added?

I would like for an additional flag on the kube api server for a whitelist of ports `--service-node-port-whitelist` that will most likely accept a comma-delimited list of port numbers which a `NodePort` service can use in addition to the already existing flag `--service-node-port-range` where a port range can be specified.

### Why is this needed?

I think its best for me to mention the use case as to why this feature would be great to add.

If you think about operating Kubernetes in an environment that doesn't have the nice integrations with CSPs, most particularly creation of `LoadBalancer` type services, you are sort of limited for inbound options on standard ports (80, 443, 3306. etc) into your cluster.

There is `NodePort` services you can use, and it seems that the default port range is `30000-32767` which is an option specified by the kube apiserver. You can modify the default port range via the flag `--service-node-port-range` to be something like `80-32767` so you can do standard `http` or `https` networking into the cluster, but this seems a bit clunky since most of the times I just want a list of the standard ports to be open and not the whole range.

So if what I am proposing is implemented you can imagine specifying the flags like so:

```
--service-node-port-whitelist=80,443,3306
```

and you wouldn't need to modify the `--service-node-port-range` and keep it at default (if you do want to provision `NodePort` services for non-standard ports).

**Comment by pranav-pandey0804** [comment]
This seems like a useful addition, especially for environments where a full port range isn't needed but certain standard ports should remain open.

**Comment by pranav-pandey0804** [comment]
Implementing a port whitelist would allow for more fine-grained control, which would enhance security while keeping configurations simple, and it's simpler than managing the whole range.

**Comment by pranav-pandey0804** [comment]
Adding a whitelist would narrow that down to just the essentials, keeping things more secure and easier to manage.

**Comment by pranav-pandey0804** [comment]
In situations where you’re working without load balancers or complex setups—like edge deployments—this would be a perfect solution. The flexibility to open just the specific ports you need would really simplify configuration.

**Comment by pranav-pandey0804** [comment]
**One thing to think about**:
 how would this new whitelist option play with existing firewall rules?

**Comment by pranav-pandey0804** [comment]
Would the kube-apiserver handle everything, or would we need to make updates to kube-proxy or similar components for the feature to work smoothly?

**Comment by aojea** [comment]
If you do that your applications in your host are going to overlap with the node ports and you'll end with chances of malicious apps hijacking your apps. 

If you have API server per example listen in A:443 , then you can not setup a nodeport in that port , same for other things.

Have you consider using hostPorts?
In addition to that, you can use loadbalancers implementations like MetalLB and similar

**Comment by yquansah** [comment]
@aojea Yeah I understand the `LoadBalancer` point especially for http(s) or http traffic, but I think the issue is a bit more general.

For example, if I *just* want to open up port 8080 for `NodePort` services, currently on the kube-apiserver I would have to specify a range. lets say I go with range `8080-40000`. I just configured a whole bunch of ports `8081`, `8082`, etc. that I do not forsee myself using for `NodePort` services.

**Comment by yquansah** [comment]
Our specific use case is that we did not want to integrate with `MetalLB` or use a `LoadBalancer` service type integrated with a CSP. We are using `microk8s` on an EC2 instance and just want to allow http(s) traffic into the cluster using a `NodePort` service, but like I stated above the issue can be way more generalized.

**Comment by yquansah** [comment]
@pranav-pandey0804 
> In situations where you’re working without load balancers or complex setups—like edge deployments—this would be a perfect solution. The flexibility to open just the specific ports you need would really simplify configuration.

Exactly, where you do not have access to LBs which _kind of_ was my use case.

**Comment by yquansah** [comment]
@aojea 
> If you do that your applications in your host are going to overlap with the node ports and you'll end with chances of malicious apps hijacking your apps.

For this point, I would say this is the responsibility of the operator of the cluster right? They should be aware about applications running on the host already (especially if you are self-hosting).

**Comment by condaatje** [comment]
perhaps there is an RFC-defined format?

this type of string looks familiar
```
80,443,3306,30000-40000
```

**Comment by aojea** [comment]
> We are using microk8s on an EC2 instance and just want to allow http(s) traffic into the cluster using a NodePort service, but like I stated above the issue can be way more generalized.

this is not a very strong use case honestly, this is not a prod cluster, right?
  we have this in kind solved, and it can be solved easily with port forwarding or any other network tricks

**Comment by yquansah** [comment]
@thockin I see. Yeah I guess my use case could be argued any way whether it fits what I'm proposing or not. But I am really trying to just point out a larger problem with NodePort pool allocation.

Maybe the existing flag on the apiserver could be used for the allocation pool and this additional flag I am proposing could be used if you want to explicitly specify a NodePort on the Service definition.

Currently, it I want to open up port 8080 for allocation I would have to specify a big range.

But like you said maybe NodePort(s) just aren't a fit for that in the first place.

**Comment by thockin** [comment]
How are clients discovering the IPs to which to speak?

**Comment by yquansah** [comment]
We use our own DNS name. To be honest, we are mimicking a LoadBalancer service type, but we just didn't want to go the route of using a LB service type.

**Comment by thockin** [comment]
And that DNS name resolves to ...?  Is it the list of node IPs?  A specific node IP?  A VIP?

**Comment by yquansah** [comment]
Yeah the list of NodeIPs

**Comment by thockin** [comment]
And you CAN'T use a service or just don't want to?

What I'm getting at is that NodePort doesn't do what you are thinking. It's just software so it could evolve, but I'm not convinced that all existing options have been explored. 

A DaemonSet with HostPort seems like it would do what you need, and be more reliable overall.  Or any sort of VIP could be handled more transparently.

**Comment by yquansah** [comment]
@thockin We can opt for the `DaemonSet` with a `HostPort`. I believe that would actually work well for our case. But your overall point now makes sense to me (NodePorts aren't the right fit for what I am thinking).

**Comment by thockin** [comment]
Great.  I'll close this for now.
